// datocms-client/lib/dump/dump.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = dump;

var _path = require("path");

var _denodeify = _interopRequireDefault(require("denodeify"));

var _rimraf = _interopRequireDefault(require("rimraf"));

var _Loader = _interopRequireDefault(require("../local/Loader"));

var _createPost2 = _interopRequireDefault(require("./createPost"));

var _createDataFile2 = _interopRequireDefault(require("./createDataFile"));

var _addToDataFile2 = _interopRequireDefault(require("./addToDataFile"));

var _i18n = _interopRequireDefault(require("../utils/i18n"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

var rimraf = (0, _denodeify.default)(_rimraf.default);
var createDirectory;

function collectOperations(base, config) {
  var operations = [];
  var dsl = {
    directory: function directory(dir, subConfig) {
      operations.push(createDirectory((0, _path.resolve)(base, dir), subConfig));
    },
    createDataFile: function createDataFile(file, format, data) {
      operations.push(function () {
        return Promise.resolve(data).then(function (dr) {
          return _createDataFile2.default.bind(null, (0, _path.resolve)(base, file), format, dr)();
        });
      });
    },
    createPost: function createPost(file, format, data) {
      operations.push(function () {
        return Promise.resolve(data).then(function (dr) {
          return _createPost2.default.bind(null, (0, _path.resolve)(base, file), format, dr)();
        });
      });
    },
    addToDataFile: function addToDataFile(file, format, data) {
      operations.push(function () {
        return Promise.resolve(data).then(function (dr) {
          return _addToDataFile2.default.bind(null, (0, _path.resolve)(base, file), format, dr)();
        });
      });
    }
  };
  config(dsl, _i18n.default);
  return operations;
}

createDirectory = function createDirectory(dir, config) {
  var operations = collectOperations(dir, config);
  return function () {
    return rimraf((0, _path.resolve)(dir, '*')).then(function () {
      return Promise.all(operations.map(function (o) {
        return o();
      }));
    }).then(function (descriptions) {
      var _ref;

      var description = "Created ".concat((0, _path.relative)(process.cwd(), dir));
      return (_ref = []).concat.apply(_ref, [description].concat(_toConsumableArray(descriptions)));
    });
  };
};

function start(path, config) {
  var operations = collectOperations(path, config);
  return function () {
    return Promise.all(operations.map(function (o) {
      return o();
    })).then(function (descriptions) {
      var _ref2;

      return (_ref2 = []).concat.apply(_ref2, _toConsumableArray(descriptions));
    });
  };
}

function dump(_x, _x2, _x3) {
  return _dump.apply(this, arguments);
}

function _dump() {
  _dump = _asyncToGenerator(
  /*#__PURE__*/
  regeneratorRuntime.mark(function _callee(configFile, client, previewMode) {
    var destinationPath,
        config,
        loader,
        _i18n$availableLocale,
        startOperation,
        _args = arguments;

    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            destinationPath = _args.length > 3 && _args[3] !== undefined ? _args[3] : process.cwd();

            /* eslint-disable global-require, import/no-dynamic-require */
            delete require.cache[configFile];
            config = require(configFile);
            /* eslint-enable global-require, import/no-dynamic-require */

            loader = new _Loader.default(client, previewMode);
            _context.next = 6;
            return loader.load();

          case 6:
            _i18n.default.availableLocales = loader.itemsRepo.site.locales;
            _i18n$availableLocale = _slicedToArray(_i18n.default.availableLocales, 1);
            _i18n.default.locale = _i18n$availableLocale[0];
            startOperation = start(destinationPath, config.bind(config, loader.itemsRepo));
            return _context.abrupt("return", startOperation());

          case 11:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, this);
  }));
  return _dump.apply(this, arguments);
}
